### ISSUE TITLE
Unblock and roll forward "Always inline IIFEs" change

### ISSUE THREAD

**Issue by brad4d** [google/closure-compiler#2409]
PR #2279 got rolled back due to a failing test within Google.

This issue tracks diagnosing the cause, fixing it and rolling forward the change from #2279.

**Comment by brad4d** [comment]
Test passes when running in debug mode.

**Comment by brad4d** [comment]
Got the test to fail in debug mode.
It definitely looks like we're removing a function defined like this:

```javascript
// No goog.module() or goog.provide()
TestHelpers = {};
(function(scope, global) {
  'use strict';

  global.TestHelpers = scope;
  global.thisGetsRemoved = function() {}
}(TestHelpers, this);
```
No, I don't know why it's defined this way, but it is in a directory tree named 'polymer',
so I expect that's related.

Clearly the author expects a non-strict environment,
since `this` is passed to the IIFE, which contains the `'use strict';` directive itself.

**Comment by brad4d** [comment]
Ah, yes. Here's an external link to the problem code.
https://github.com/PolymerElements/iron-test-helpers/blob/master/test-helpers.js

**Comment by ChadKillingsworth** [comment]
So the hard part to figure out is this getting removed by later passes because we get better optimizations or is this actually caused by the inlining?

An easy fix might be to not to force inlining if `this` is passed as an argument.

**Comment by brad4d** [comment]
I reproduced this using `--print_source_after_each_pass` and determined that the guilty party is "smartNamePass".
Its output differs from the previous pass only in that this chunk of code is removed.

```js
var scope$jscomp$inline_2690 = {};
var global$jscomp$inline_2691 = this;
global$jscomp$inline_2691.TestHelpers = scope$jscomp$inline_2690;
global$jscomp$inline_2691.flushAsynchronousOperations = function() {
  Polymer.dom.flush();
  window.CustomElements && window.CustomElements.takeRecords();
};
global$jscomp$inline_2691.forceXIfStamp = function(node$jscomp$inline_2692) {
  var templates$jscomp$inline_2693 = Polymer.dom(node$jscomp$inline_2692.root).querySelectorAll("template[is=dom-if]");
  var tmpl$jscomp$inline_2694;
  var i$jscomp$inline_2695 = 0;
  for (;tmpl$jscomp$inline_2694 = templates$jscomp$inline_2693[i$jscomp$inline_2695];i$jscomp$inline_2695++) {
    tmpl$jscomp$inline_2694.render();
  }
  global$jscomp$inline_2691.flushAsynchronousOperations();
};
global$jscomp$inline_2691.fireEvent = function(type$jscomp$inline_2696, props$jscomp$inline_2697, node$jscomp$inline_2698) {
  var event$jscomp$inline_2699 = new CustomEvent(type$jscomp$inline_2696, {bubbles:true, cancelable:true});
  var p$jscomp$inline_2700;
  for (p$jscomp$inline_2700 in props$jscomp$inline_2697) {
    event$jscomp$inline_2699[p$jscomp$inline_2700] = props$jscomp$inline_2697[p$jscomp$inline_2700];
  }
  node$jscomp$inline_2698.dispatchEvent(event$jscomp$inline_2699);
};
global$jscomp$inline_2691.skipUnless = function(condition$jscomp$inline_2701, test$jscomp$inline_2702) {
  var isAsyncTest$jscomp$inline_2703 = !!test$jscomp$inline_2702.length;
  return function(done$jscomp$inline_2704) {
    if (!condition$jscomp$inline_2701()) {
      return done$jscomp$inline_2704();
    }
    var result$jscomp$inline_2705 = test$jscomp$inline_2702.call(this, done$jscomp$inline_2704);
    isAsyncTest$jscomp$inline_2703 || done$jscomp$inline_2704();
    return result$jscomp$inline_2705;
  };
};
scope$jscomp$inline_2690.flushAsynchronousOperations = global$jscomp$inline_2691.flushAsynchronousOperations;
scope$jscomp$inline_2690.forceXIfStamp = global$jscomp$inline_2691.forceXIfStamp;
scope$jscomp$inline_2690.fireEvent = global$jscomp$inline_2691.fireEvent;
scope$jscomp$inline_2690.skipUnless = global$jscomp$inline_2691.skipUnless;
```

**Comment by ChadKillingsworth** [comment]
Well now I have a test case. I'll see if I can work on smart name removal.

**Comment by brad4d** [comment]
I pared the test case down even further, and it seems that smart name removal isn't the only potential problem.

```javascript
/ parseInputs yields:
// ************************************
(function(global) {
  global.skipUnless = function() {
  };
})(this);


// inlineFunctions yields:
// ************************************
 {
  this.skipUnless = function() {
  };
}
;

// removeUnusedClassProperties yields:
// ************************************
 {
  (function() {
  });
}
;

// peepholeOptimizations yields:
// ************************************

```

The above run was done with the "always inline IIFEs" logic put back in place.
Without that change, the result looks like this.

```javascript
// parseInputs yields:
// ************************************
(function(global) {
  global.skipUnless = function() {
  };
})(this);


// renameProperties yields:
// ************************************
(function(global) {
  global.$skipUnless$ = function() {
  };
})(this);


// renameVars yields:
// ************************************
(function($global$$) {
  $global$$.$skipUnless$ = function() {
  };
})(this);

```

**Comment by brad4d** [comment]
In summary:
It seems that at least 2 passes fail to recognize `this.something = ...` in the global scope as a global symbol definition, smartNamePass and removeUnusedClassProperties.

I think we have 2 options for a fix.
1. Change the IIFE inlining behavior to avoid inlining if the IIFE is called with `this` as one of its arguments.
2. Fix (at least) smartNamePass and removeUnusedClassProperties to recognize assignments to `this.something = ...` in the global scope as global definitions and not remove them.

I think 1 is easiest and safest, though 2 would actually put us in a better state, because the compiler would then be better at recognizing types defined in this manner.

**Comment by brad4d** [comment]
Discussed this with @dimvar 
We've agreed to go with the quick fix of avoiding function inlining when we're passing `this` to the IIFE in global scope.

**Comment by ChadKillingsworth** [comment]
I'll try to tackle the other two passes sometime. It would be nice to have this handled. But I'd rather see IIFE inlining landed.

**Comment by brad4d** [comment]
@ChadKillingsworth 
I want to do the fix, but first I wrote some tests for it.
See
https://github.com/google/closure-compiler/pull/2447/files

Crazy thing is the `testKeepIifeWithThisArgInGlobalScope()` test is failing already without rolling forward the "always inline IIFEs change". The global-scope IIFE is getting inlined in the unit test, but when I run the compiler on this repro, it is not.

```js
(function(global) {
  global.skipUnless = function() {};
})(this);
```
Do you know what I'm missing here? Thanks.

**Comment by ChadKillingsworth** [comment]
Off the top of my head, no. I'm on vacation and didn't bring my laptop or I
would check.

**Comment by brad4d** [comment]
Now that `smartNamePass` is gone, it may be worth trying to do this again.

**Comment by brad4d** [comment]
@ChadKillingsworth is the "always inline IIFEs" change still something you want?
